### ISSUE TITLE
Question: deep network fails to learn accurately

### ISSUE THREAD

**Issue by alexge233** [OpenANN/OpenANN#174]
I realize this question is wide and broad, I'm just looking for some hints.

I'm training a network (3 layers, 100, 50, 20 nodes) to classify relationships between POS tags.
The input is a sparsely-coded vector where each node is activated in the presence of a POS tag, in connection to the presence of a POS tag (e.g., for 32 POS tags, I have a vector of 64 input nodes).

I've tried with various hidden layers, hidden neurons, added drop-out, L2 and L1 regularization, but in every case the best accuracy I am able to achieve is around 72%, using MBSGD. The hidden activation is a ReLU and the output is a Softmax. Cross Entropy is always very large, regardless of the training epochs.

Any hints on what it is that may be causing this? My training samples, my design or something else?
Other similarly sparsely-coded experiments (not using POS tags) achieve high accuracy (99.75%) and work quite well.

**Comment by AlexanderFabisch** [comment]
I assume that 72% is the accuracy on the training set and your model is underfitting. There are many possible reasons. Maybe the most likely reasons are

* your model is not complex enough -> add hidden layers or nodes
* your objective function is too difficult to optimize -> try conjugate gradient or L-BFGS
* your learning rate is too large -> try smaller values, set the momentum to 0, no learning rate decay etc.

What is the training accuracy with a very simple model like k-nearest neighbors? When you not sure which architecture to choose for your neural net you should start without any hidden layers and then add one hidden layer and increase the number of nodes step by step. When you are sure that you can reach a good accuracy, you can decrease the number of neurons and add more hidden layers. That's how I would do this.

Dropout does not decrease your training error. It can be used to increase the accuracy on your test set by 1-2%.

**Comment by alexge233** [comment]
@AlexanderFabisch Many thanks for the hints!

1. I've tried adding more hidden layers or nodes, up to a point I got a small increase, then it dropped again.
2. I will try another training algorithm.
3. How can I set the learning rate? I'm using a `OpenANN::StoppingCriteria` setting epochs and minimal value difference. I was under the impression that the learning rate was auto-decremented?

I will follow your advice and try to go from simple to complex design.
Many thanks!

**Comment by AlexanderFabisch** [comment]
You can set the learning rate etc. in the [constructor of MBSGD](http://openann.github.io/OpenANN-apidoc/classOpenANN_1_1MBSGD.html#af5f366446afcd555ad69c2f49a8ae979), see [here](https://github.com/OpenANN/OpenANN/blob/master/benchmarks/sarcos/benchmark.py#L111) for an example.

**Comment by alexge233** [comment]
@AlexanderFabisch Thanks again for your help.
Im still a wee bit confused on how to use it. AFAIK, I have a training call:

```
auto net = std::unique_ptr<OpenANN::Net>(new OpenANN::Net());
OpenANN::makeMLNN(*net, OpenANN::RECTIFIER, OpenANN::SOFTMAX, input, output, 1, 200);
net->trainingSet(dataset);

OpenANN::StoppingCriteria stop;
stop.maximalIterations = 5000;
stop.minimalValueDifferences = 0.001;
OpenANN::train(*net, "MBSGD", OpenANN::CE, stop);
```

I can't seem to find the documentation which shows how to use a custom `MBSGD` object for training.
The `OpenANN::train` method seems to only accept an `std::string` parameter (http://openann.github.io/OpenANN-apidoc/namespaceOpenANN.html#a5fcdd987e4be42f5de6109e26c9691a5) is there some way I can directly use the `MBSGD` object?

**Comment by AlexanderFabisch** [comment]
`OpenANN::train` is a simpilfied interface that you cannot use in this case. You can find examples here:

* [DataStream](https://github.com/OpenANN/OpenANN/blob/master/benchmarks/mnist/mnist.cpp#L84)
* [optimizer interface](https://github.com/OpenANN/OpenANN/blob/master/benchmarks/cifar10/cifar10.cpp#L55)

**Comment by alexge233** [comment]
Many thanks, I think the problem is with the samples, and not the training.
In any case, is there a publication I can cite/reference?

**Comment by AlexanderFabisch** [comment]
About OpenANN? Not really. It has been developed mainly for [this publication](http://www.informatik.uni-bremen.de/~afabisch/files/2013_NN_LCS.pdf). You can just add a link if you want.
